Specialize multi-arg + for Array

martinholters opened the pull request:
... to avoid unnecessary creation of temporaries by repeatedly doing
two-arg addition.

stevengj commented:
Do we have a test case for 3-arg array `+`?

martinholters commented:
> Do we have a test case for 3-arg array `+`?

No, it is tested indirectly as part of matrix exponential and trigonometric functions only. It probably makes sense to add a test here.

But I realized there is a more severe caveat: Any `AbstractArray` subtypes specializing two-arg `+` but not multi-arg `+` may now suddenly behave differently. E.g. `+(::Diagonal, ::Diagonal)` returns a `Diagonal`, but the multi-arg fallback from this PR would make `+(::Diagonal, ::Diagonal, ::Diagonal)` return a `SparseMatrixCSC` (as the result of the broadcast). The long term goal should probably be to specialize `+(::Diagonal...)` (and likewise for other `AbstractArray` subtypes). While we can do that in base, it's unclear what the fall-out in other packages due to this PR would be, so more defensively, we could _only_ do this for strict subtypes of `AbstractArray` for now. Opinions? (Marking as WIP until this has been resolved.)

stevengj commented:
Simpler workaround use the `broadcast` method only for `Array`, for now.  That's probably the main practical case where this matters in real code, anyway.

(If someone is adding a whole bunch of `Diagonal` matrices together, I'm guessing that is not a performance-limiting part of their code.)

martinholters committed:
Specialize multi-arg + for Array

... to avoid unnecessary creation of temporaries by repeatedly doing
two-arg addition.

martinholters commented:
> Simpler workaround use the fused broadcast method only for `Array`, for now.

Yes, done that now.

I've also added a very basic test. I wanted to add multi-arg cases where two-arg `+` was being tested ... but failed to find anything besides the basic test at the top of `arrayops.jl`. Any pointers?

martinholters commented:
AV failure looks unrelated:
```
	From worker 9:	Exception: EXCEPTION_ACCESS_VIOLATION at 0x680de1a4 -- dereferenceable_size at /cygdrive/c/projects/julia/src/cygdrive/c/projects/julia/src\cgutils.cpp:338
	From worker 9:	in expression starting at C:\projects\julia\julia-ac312fc8b4\share\julia\test\sparse\higherorderfns.jl:270
	From worker 9:	jl_is_array_type at /cygdrive/c/projects/julia/src/cygdrive/c/projects/julia/src\julia.h:955 [inlined]
	From worker 9:	dereferenceable_size at /cygdrive/c/projects/julia/src/cygdrive/c/projects/julia/src\cgutils.cpp:335
	From worker 9:	maybe_mark_load_dereferenceable at /cygdrive/c/projects/julia/src/cygdrive/c/projects/julia/src\cgutils.cpp:381 [inlined]
	From worker 9:	emit_local at /cygdrive/c/projects/julia/src/cygdrive/c/projects/julia/src\codegen.cpp:3346 [inlined]
```
Already known?

martinholters commented:
Ref. https://github.com/JuliaLang/julia/pull/24283#issuecomment-341718957 concerning AV failure.
